Mathlib/RingTheory/Norm/Transitivity.lean

```diff
@@ -0,0 +1,198 @@
+
+lemma eval_zero_det_det : eval 0 (f.polyToMatrix (cornerAddX M k).det).det = (f M.det).det := by
+  rw [← coe_evalRingHom, RingHom.map_det, ← RingHom.comp_apply,
+    evalRingHom_mapMatrix_comp_polyToMatrix, f.comp_apply, RingHom.map_det]
+  congr; ext; simp [cornerAddX, diagonal, apply_ite]
+
+lemma eval_zero_compRingEquiv_det :
+    eval 0 (compRingEquiv m n R[X] <| f.polyToMatrix.mapMatrix <| cornerAddX M k).det =
+      (compRingEquiv m n R <| f.mapMatrix M).det := by
+  simp_rw [← coe_evalRingHom, RingHom.map_det, ← RingEquiv.coe_toRingHom, ← RingHom.comp_apply,
+    ← RingHom.comp_assoc, evalRingHom_mapMatrix_comp_compRingEquiv, RingHom.comp_assoc,
+    RingHom.mapMatrix_comp, evalRingHom_mapMatrix_comp_polyToMatrix, ← RingHom.mapMatrix_comp,
+    RingHom.comp_apply]
+  congr with i j
+  simp [cornerAddX, diagonal, apply_ite]
+
+theorem compRingEquiv_det_mul_pow :
+    ((f.mapMatrix M).compRingEquiv m n R).det * (f (M k k)).det ^ (Fintype.card m - 1) =
+      (f (M k k)).det * ((f.mapMatrix (mulAuxMatBlock M k)).compRingEquiv _ n R).det := by
+  trans ((f.mapMatrix (M * auxMat M k)).compRingEquiv m n R).det
+  · simp_rw [_root_.map_mul, det_mul, ((auxMat_blockTriangular M k).mapMatrix f).compRingEquiv
+      |>.det_fintype, Fintype.prod_Prop, compRingEquiv_apply, comp_toSquareBlock (b := (· ≠ k)),
+      det_reindex_self, f.mapMatrix_apply, toSquareBlock_map, auxMat_toSquareBlock_true,
+      auxMat_toSquareBlock_false, smul_one_eq_diagonal, ← diagonal_one,
+      diagonal_map (map_zero _), comp_diagonal, det_reindex_self]
+    simp
+  · simp_rw [((mul_auxMat_blockTriangular M k).mapMatrix f).compRingEquiv.det_fintype,
+      Fintype.prod_Prop, compRingEquiv_apply, comp_toSquareBlock (b := (· = k)), det_reindex_self,
+      f.mapMatrix_apply, toSquareBlock_map, mul_auxMat_toSquareBlock_true, smul_one_eq_diagonal,
+      diagonal_map (map_zero _), comp_diagonal, det_reindex_self]
+    simp [mulAuxMatBlock]
+
+variable {M f} in
+lemma det_det_aux
+    (ih : ∀ M, (f M.det).det = ((f.mapMatrix M).compRingEquiv {a // (a = k) = False} n R).det) :
+    ((f M.det).det - ((f.mapMatrix M).compRingEquiv m n R).det) *
+      (f (M k k)).det ^ (Fintype.card m - 1) = 0 := by
+  rw [sub_mul, compRingEquiv_det_mul_pow, ← det_pow, ← map_pow, ← det_mul, ← _root_.map_mul,
+    det_mul_corner_pow, _root_.map_mul, det_mul, mulAuxMatBlock, ih, sub_self]
+
+end Algebra.Norm.Transitivity
+
+open Algebra.Norm.Transitivity
+
+/-- The main result in Silvester's paper *Determinants of Block Matrices*: the determinant of
+a block matrix with commuting, equal-sized, square blocks can be computed by taking two
+determinants in a row: first take the determinant over the commutative ring generated by the
+blocks (`S` here), then take the determinant over the base ring. -/
+theorem Matrix.det_det [Fintype m] [Fintype n] (f : S →+* Matrix n n R) :
+    (f M.det).det = ((f.mapMatrix M).compRingEquiv m n R).det := by
```

Comment:
Should this be stated as
```suggestion
    (f M.det).det = ((M.map f).comp m m n n R).det := by
```
or is this spelling just going to be annoying?

Reply:
Not hard to fix at all with the `comp` spelling! I think we ought to do it because `comp` is the simp normal form.

Reply:
What about `map` vs `mapMatrix`? Or is that one less clear?

Reply:
Oh I missed that. Will submit another PR ...

Reply:
#20009